// @aws-cdk/aws-route53/lib/hosted-zone-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cdk = require("@aws-cdk/cdk");
const cxapi = require("@aws-cdk/cx-api");
const hosted_zone_1 = require("./hosted-zone");
const DEFAULT_HOSTED_ZONE = {
    Id: '/hostedzone/DUMMY',
    Name: 'example.com',
};
/**
 * Context provider that will lookup the Hosted Zone ID for the given arguments
 */
class HostedZoneProvider {
    constructor(context, props) {
        this.provider = new cdk.ContextProvider(context, cxapi.HOSTED_ZONE_PROVIDER, props);
    }
    /**
     * This method calls `findHostedZone` and returns the imported hosted zone
     */
    findAndImport(scope, id) {
        return hosted_zone_1.HostedZone.fromHostedZoneAttributes(scope, id, this.findHostedZone());
    }
    /**
     * Return the hosted zone meeting the filter
     */
    findHostedZone() {
        const zone = this.provider.getValue(DEFAULT_HOSTED_ZONE);
        // CDK handles the '.' at the end, so remove it here
        if (zone.Name.endsWith('.')) {
            zone.Name = zone.Name.substring(0, zone.Name.length - 1);
        }
        return {
            hostedZoneId: zone.Id,
            zoneName: zone.Name,
        };
    }
}
exports.HostedZoneProvider = HostedZoneProvider;
