// @aws-cdk/aws-elasticloadbalancingv2/lib/shared/base-listener.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cdk_1 = require("@aws-cdk/cdk");
const elasticloadbalancingv2_generated_1 = require("../elasticloadbalancingv2.generated");
/**
 * Base class for listeners
 */
class BaseListener extends cdk_1.Resource {
    constructor(scope, id, additionalProps) {
        super(scope, id);
        this.defaultActions = [];
        const resource = new elasticloadbalancingv2_generated_1.CfnListener(this, 'Resource', {
            ...additionalProps,
            defaultActions: cdk_1.Lazy.anyValue({ produce: () => this.defaultActions }),
        });
        this.listenerArn = resource.refAsString;
    }
    /**
     * Validate this listener
     */
    validate() {
        if (this.defaultActions.length === 0) {
            return ['Listener needs at least one default target group (call addTargetGroups)'];
        }
        return [];
    }
    /**
     * Add an action to the list of default actions of this listener
     * @internal
     */
    _addDefaultAction(action) {
        this.defaultActions.push(action);
    }
    /**
     * Add a TargetGroup to the list of default actions of this listener
     * @internal
     */
    _addDefaultTargetGroup(targetGroup) {
        this._addDefaultAction({
            targetGroupArn: targetGroup.targetGroupArn,
            type: 'forward'
        });
    }
}
exports.BaseListener = BaseListener;
