// @aws-cdk/cdk/lib/physical-name.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const lazy_1 = require("./lazy");
const physical_name_generator_1 = require("./physical-name-generator");
/**
 * Represents the physical (that is, visible in the AWS Console) name of a resource.
 * These can be known at code runtime,
 * in the case of fixed, customer-provided names,
 * at synthesis time, in case of automatically assigned names by the framework,
 * or only at deploy time, assigned by CloudFormation -
 * which is the default.
 *
 * @see #auto()
 * @see #of()
 */
class PhysicalName {
    /**
     * A physical name that will be automatically generated.
     * The details of how and when that generation will occur can be customized by passing an instance
     * of {@link AutoPhysicalNameOptions}.
     * By default, the physical name of the resource will be generated by CloudFormation,
     * at deploy time.
     */
    static auto(options = {}) {
        return new LateBoundPhysicalName(options);
    }
    /**
     * A fixed physical name (one that is known statically, at synthesis time).
     *
     * @param name the name to assign
     */
    static of(name) {
        return new FixedPhysicalName(name);
    }
    constructor() {
    }
    /**
     * A callback method called if the resource this physical name
     * belongs to has been used in a cross-environment fashion.
     * Allows late-binding of the actual name to only be generated if necessary,
     * and kept deploy-time if not.
     *
     * The default implementation does nothing,
     * but can be overridden in subclasses.
     *
     * @param _resource the resource this physical name belongs to
     * @internal
     */
    _resolveCrossEnvironment(_resource) {
        // does nothing
    }
}
exports.PhysicalName = PhysicalName;
// implementations are private to this module,
// we only surface them through static factory methods
class FixedPhysicalName extends PhysicalName {
    constructor(value) {
        super();
        this.value = value;
    }
}
class LateBoundPhysicalName extends PhysicalName {
    constructor(options) {
        super();
        this.value = options.crossEnvironment
            ? lazy_1.Lazy.stringValue({ produce: () => this.name })
            : undefined;
    }
    /** @internal */
    _resolveCrossEnvironment(resource) {
        if (!this.value) {
            // error out - a deploy-time name cannot be used across environments
            throw new Error(`Cannot use resource '${resource.node.path}' in a cross-environment fashion, ` +
                "as it doesn't have a physical name set. Use PhysicalName.auto({ crossEnvironment: true }) to enable cross env name allocation");
        }
        if (!this.name) {
            this.name = physical_name_generator_1.generatePhysicalName(resource);
        }
    }
}
