// @aws-cdk/cdk/lib/private/cross-environment-token.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const stack_1 = require("../stack");
const stack_trace_1 = require("../stack-trace");
/**
 * A Token that represents a reference that spans accounts and/or regions,
 * and so requires the resources to have physical names.
 * You should never need to interact with these directly,
 * instead use the {@link ResourceIdentifiers} class.
 * This class is private to the @aws-cdk/cdk package.
 */
class CrossEnvironmentToken {
    /**
     * @param regularValue the value used when this is referenced NOT from a cross account and/or region Stack
     * @param crossEnvironmentValue the value used when this is referenced from a cross account and/or region Stack
     * @param resource the scope this reference is mastered in. Used to determine the owning Stack
     * @param displayName a short name to be used in Token display
     */
    constructor(regularValue, crossEnvironmentValue, resource) {
        this.regularValue = regularValue;
        this.crossEnvironmentValue = crossEnvironmentValue;
        this.resource = resource;
        this.resource = resource;
        this.creationStack = stack_trace_1.captureStackTrace();
    }
    resolve(context) {
        const consumingStack = stack_1.Stack.of(context.scope);
        const owningStack = stack_1.Stack.of(this.resource);
        if (consumingStack.environment !== owningStack.environment) {
            this.resource.physicalName._resolveCrossEnvironment(this.resource);
            return this.crossEnvironmentValue;
        }
        else {
            return this.regularValue;
        }
    }
}
exports.CrossEnvironmentToken = CrossEnvironmentToken;
class CrossEnvironmentPhysicalArnToken extends CrossEnvironmentToken {
    constructor(regularValue, arnComponents, resource) {
        super(regularValue, stack_1.Stack.of(resource).formatArn(arnComponents), resource);
    }
}
exports.CrossEnvironmentPhysicalArnToken = CrossEnvironmentPhysicalArnToken;
class CrossEnvironmentPhysicalNameToken extends CrossEnvironmentToken {
    constructor(regularValue, resource) {
        super(regularValue, resource.physicalName.value, resource);
    }
}
exports.CrossEnvironmentPhysicalNameToken = CrossEnvironmentPhysicalNameToken;
